community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/codegen/ir/Instruction.scala

```diff
@@ -25,42 +25,36 @@ trait Instruction {
   // Actual code produced by element
   def generateCode(): String
 
-  final def importedClasses() = allLeafs.flatMap(_._importedClasses()).toSet
+  //Initialises necessary data-structures. Is inserted at the top of the generated method
+  def generateInit(): String
 
   //generate class level members
   def members(): String
 
-  //Initialises necessary data-structures. Is inserted at the top of the generated method
-  def generateInit(): String
+  protected def children: Seq[Instruction] = Seq.empty
 
-  def methods: Seq[Method] = {
-    (allLeafs :+ this).flatMap(_._method)
+  private def treeView: Seq[Instruction] = {
+    children.foldLeft(this +: children) { case (acc, child) => acc ++ child.treeView
+    }
   }
 
-  def exceptions: Set[ExceptionCodeGen] = allLeafs.flatMap(_._exceptions()).toSet
+  // Aggregating methods -- final to prevent overriding
+  final def allImportedClasses: Set[String] = treeView.flatMap(_.localImportedClasses).toSet
 
-  // Generates import list for class - implement this!
-  protected def _importedClasses(): Set[String]
+  final def allMethods: Seq[Method] = treeView.flatMap(_.localMethod)
 
-  protected def _exceptions(): Set[ExceptionCodeGen] = Set.empty
+  final def allExceptions: Set[ExceptionCodeGen] = treeView.flatMap(_.localExceptions).toSet
 
-  protected def children: Seq[Instruction] = Seq.empty
+  final def allOperatorIds: Set[String] = treeView.flatMap(_.localOperatorId).toSet
 
-  def operatorId: Option[String] = None
+  // Override these
+  protected def localImportedClasses: Set[String] = Set.empty
```

Comment:
What is the gain of having `localX` and `X` why not just remove the former and have `allX` operate directly on `X`.